src/CoreWCF.Primitives/src/CoreWCF/Channels/TransportSecurityBindingElement.cs

```diff
@@ -0,0 +1,216 @@
+                if (this.EndpointSupportingTokenParameters.Endorsing.Count > 0)
+                    scParameters = this.EndpointSupportingTokenParameters.Endorsing[0] as SecureConversationSecurityTokenParameters;
+                if (scParameters != null)
+                    return scParameters.RequireCancellation;
+                else
+                    return false;
+            }
+        }
+
+        internal override bool SupportsDuplex
+        {
+            get { return true; }
+        }
+
+        internal override bool SupportsRequestReply
+        {
+            get { return true; }
+        }
+
+
+        internal override SecurityProtocolFactory CreateSecurityProtocolFactory<TChannel>(BindingContext context, SecurityCredentialsManager credentialsManager, bool isForService, BindingContext issuerBindingContext)
+        {
+            if (context == null)
+                throw DiagnosticUtility.ExceptionUtility.ThrowHelperArgumentNull("context");
+            if (credentialsManager == null)
+                throw DiagnosticUtility.ExceptionUtility.ThrowHelperArgumentNull("credentialsManager");
+
+            TransportSecurityProtocolFactory protocolFactory = new TransportSecurityProtocolFactory();
+            // if (isForService)
+            //     base.ApplyAuditBehaviorSettings(context, protocolFactory);
+            base.ConfigureProtocolFactory(protocolFactory, credentialsManager, isForService, issuerBindingContext, context.Binding);
+            protocolFactory.DetectReplays = false;
+
+            return protocolFactory;
+        }
+
+        protected override IServiceDispatcher BuildServiceDispatcherCore<TChannel>(BindingContext context, IServiceDispatcher serviceDispatcher)
+        {
+            SecurityServiceDispatcher securityServiceDispatcher = new SecurityServiceDispatcher(this, context, serviceDispatcher);
+            SecurityCredentialsManager credentialsManager = serviceDispatcher.Host.Description.Behaviors.Find<SecurityCredentialsManager>();
+            if (credentialsManager == null)
+                credentialsManager = ServiceCredentials.CreateDefaultCredentials();
+            
+            SecureConversationSecurityTokenParameters scParameters;
+            if (this.EndpointSupportingTokenParameters.Endorsing.Count > 0)
+                scParameters = this.EndpointSupportingTokenParameters.Endorsing[0] as SecureConversationSecurityTokenParameters;
+            else
+                scParameters = null;
+
+            bool requireDemuxer = false; // TODO later  RequiresChannelDemuxer();
+            ChannelBuilder channelBuilder = new ChannelBuilder(context, requireDemuxer);
+
+            if (requireDemuxer)
+            {
+                ApplyPropertiesOnDemuxer(channelBuilder, context);
+            }
+
+            BindingContext issuerBindingContext = context.Clone();
+            issuerBindingContext.BindingParameters.Add(credentialsManager);
+            if (scParameters != null)
```

Comment:
Why are you adding the credentialsManager to the BindingParameters? There might already be one in there. This shouldn't be needed.

Reply:
I am adding if it's absent.. to be consistent with WCF framework.. mostly there would be another scenario which would be needing this... but you are right, given the current feature we are porting the credential manager is already populated.